pkg/controller/kibana/config_settings.go

```diff
@@ -127,6 +135,19 @@ func NewConfigSettings(ctx context.Context, client k8s.Client, kb kbv1.Kibana, v
 	return CanonicalConfig{cfg}, nil
 }
 
+// Some previously-unsupported keys cause Kibana to error out even if the values are empty. ucfg cannot ignore fields easily so this is necessary to
+// support older versions
+func filterConfigSettings(kb kbv1.Kibana, cfg *settings.CanonicalConfig) (*settings.CanonicalConfig, error) {
+	ver, err := version.Parse(kb.Spec.Version)
+	if err != nil {
+		return cfg, err
+	}
+	if !ver.IsSameOrAfter(version.From(7, 6, 0)) {
+		_, err = (*ucfg.Config)(cfg).Remove(XpackEncryptedSavedObjects, -1, settings.Options...)
```

Comment:
ucfg supports Initialisers and with a separate struct for > 7.6.0 we could avoid the type conversion and the angling for values here but I am not sure the result is any better:

```diff
diff --git a/pkg/controller/kibana/config_settings.go b/pkg/controller/kibana/config_settings.go
index 5f14f3f4..6075691a 100644
--- a/pkg/controller/kibana/config_settings.go
+++ b/pkg/controller/kibana/config_settings.go
@@ -160,11 +160,27 @@ func VersionDefaults(kb *kbv1.Kibana, v version.Version) *settings.CanonicalConf
 
 // reusableSettings captures secrets settings in the Kibana configuration that we want to reuse.
 type reusableSettings struct {
+	EncryptionKey string `config:"xpack.security.encryptionKey"`
+	ReportingKey  string `config:"xpack.reporting.encryptionKey"`
+}
+
+func (r *reusableSettings) InitDefaults() {
+	r.EncryptionKey = string(common.RandomBytes(64))
+	r.ReportingKey = string(common.RandomBytes(64))
+}
+
+type reusableSettingsV76 struct {
 	EncryptionKey   string `config:"xpack.security.encryptionKey"`
 	ReportingKey    string `config:"xpack.reporting.encryptionKey"`
 	SavedObjectsKey string `config:"xpack.encryptedSavedObjects.encryptionKey"`
 }
 
+func (r *reusableSettingsV76) InitDefaults() {
+	r.EncryptionKey = string(common.RandomBytes(64))
+	r.ReportingKey = string(common.RandomBytes(64))
+	r.SavedObjectsKey = string(common.RandomBytes(64))
+}
+
 // getExistingConfig retrieves the canonical config for a given Kibana, if one exists
 func getExistingConfig(client k8s.Client, kb kbv1.Kibana) (*settings.CanonicalConfig, error) {
 	var secret corev1.Secret
@@ -197,18 +213,8 @@ func getOrCreateReusableSettings(c k8s.Client, kb kbv1.Kibana) (*settings.Canoni
 	if err != nil {
 		return nil, err
 	}
-
-	var r reusableSettings
 	if cfg == nil {
-		r = reusableSettings{}
-	} else if err := cfg.Unpack(&r); err != nil {
-		return nil, err
-	}
-	if len(r.EncryptionKey) == 0 {
-		r.EncryptionKey = string(common.RandomBytes(64))
-	}
-	if len(r.ReportingKey) == 0 {
-		r.ReportingKey = string(common.RandomBytes(64))
+		cfg = settings.NewCanonicalConfig()
 	}
 
 	kbVer, err := version.Parse(kb.Spec.Version)
@@ -216,8 +222,16 @@ func getOrCreateReusableSettings(c k8s.Client, kb kbv1.Kibana) (*settings.Canoni
 		return nil, err
 	}
 	// xpack.encryptedSavedObjects.encryptionKey was only added in 7.6.0 and earlier versions error out
-	if len(r.SavedObjectsKey) == 0 && kbVer.IsSameOrAfter(version.From(7, 6, 0)) {
-		r.SavedObjectsKey = string(common.RandomBytes(64))
+	if kbVer.IsSameOrAfter(version.From(7, 6, 0)) {
+		var r reusableSettingsV76
+		if err := cfg.Unpack(&r); err != nil {
+			return nil, errors.Wrap(err, "on unpack")
+		}
+		return settings.MustCanonicalConfig(r), nil
+	}
+	var r reusableSettings
+	if err := cfg.Unpack(&r); err != nil {
+		return nil, errors.Wrap(err, "on unpack")
 	}
 	return settings.MustCanonicalConfig(r), nil
 }
diff --git a/pkg/controller/kibana/config_settings_test.go b/pkg/controller/kibana/config_settings_test.go
index 33f4e970..d0f4fb84 100644
--- a/pkg/controller/kibana/config_settings_test.go
+++ b/pkg/controller/kibana/config_settings_test.go
@@ -66,6 +66,12 @@ func Test_reuseOrGenerateSecrets(t *testing.T) {
 	kb75 := mkKibana()
 	kb75.Spec.Version = "7.5.0"
 
+	type configVerifier struct {
+		EncryptionKey   string `config:"xpack.security.encryptionKey"`
+		ReportingKey    string `config:"xpack.reporting.encryptionKey"`
+		SavedObjectsKey string `config:"xpack.encryptedSavedObjects.encryptionKey"`
+	}
+
 	tests := []struct {
 		name      string
 		args      args
@@ -109,8 +115,8 @@ func Test_reuseOrGenerateSecrets(t *testing.T) {
 			},
 			assertion: func(t *testing.T, got *settings.CanonicalConfig, err error) {
 				// Unpack the configuration to check that some default reusable settings have been generated
-				var r reusableSettings
-				assert.NoError(t, got.Unpack(&r))
+				var r configVerifier
+				assert.NoError(t, got.Unpack(&r), "on unpack")
 				assert.Equal(t, len(r.EncryptionKey), 64) // key length should be 64
 				assert.Equal(t, len(r.ReportingKey), 64)
 				assert.Equal(t, len(r.SavedObjectsKey), 64)
@@ -132,11 +138,11 @@ func Test_reuseOrGenerateSecrets(t *testing.T) {
 			},
 			assertion: func(t *testing.T, got *settings.CanonicalConfig, err error) {
 				// Unpack the configuration to check that some default reusable settings have been generated
-				var r reusableSettings
+				var r configVerifier
 				assert.NoError(t, got.Unpack(&r))
 				assert.Equal(t, len(r.EncryptionKey), 64) // key length should be 64
 				assert.Equal(t, len(r.ReportingKey), 64)
-				assert.Equal(t, len(r.SavedObjectsKey), 0) // is only introduced in 7.6.0
+				assert.Equal(t, len(r.SavedObjectsKey), 0) // only introduced in 7.6.0
 			},
 		},
 	}
```

Reply:
Oh nice, I missed `InitDefaults`. I'm not sure either is really that much clearer than the other, so unless someone else has a different opinion I'll probably merge it as is.